Add support for importmap integrity

yoavweiss opened the pull request:
#### 339bcec0f9ff29fbd56e7209f4a01f2b12b14ac3
<pre>
Add support for importmap integrity
<a href="https://bugs.webkit.org/show_bug.cgi?id=272884">https://bugs.webkit.org/show_bug.cgi?id=272884</a>

Reviewed by Ryosuke Niwa.

Imported ES modules can&apos;t currently have integrity checks, which means
they can&apos;t be used in sites where integrity checks are a necessity, for
security and privacy reasons.
This implements such support, by adding an &quot;integrity&quot; section to import
maps.

See <a href="https://github.com/whatwg/html/pull/10269">https://github.com/whatwg/html/pull/10269</a>

* LayoutTests/TestExpectations: Ignored console logs to avoid flakiness
* LayoutTests/imported/w3c/web-platform-tests/import-maps/WEB_FEATURES.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resources/test-helper.js:
(createTestIframe): Updated through import.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/dynamic-integrity-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/dynamic-integrity.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity-valid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/nonimport-integrity-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/nonimport-integrity.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/static-integrity-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/static-integrity.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/w3c-import.log: Imports.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https-expected.txt: Updated.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https.html: Updated to cover Request.integrity.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/fetch-request-resources-iframe.https.html: Updated to cover Request.integrity.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https-expected.txt: Updated.
* Source/JavaScriptCore/runtime/ImportMap.cpp:
(JSC::ImportMap::resolveImportMatch): Typos and spec link.
(JSC::parseURLLikeModuleSpecifier): Typos and spec link.
(JSC::ImportMap::resolve const): Typos and spec link.
(JSC::normalizeSpecifierKey): Typos and spec link.
(JSC::sortAndNormalizeSpecifierMap): Typos and spec link.
(JSC::ImportMap::registerImportMap): Add parsing for the integrity
section.
(JSC::ImportMap::getIntegrity const): Getter for integrity based on URL.
* Source/JavaScriptCore/runtime/ImportMap.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::importModule): Add integrity to outgoing
requests.
(WebCore::ScriptModuleLoader::notifyFinished): Enforce integrity from
the importmap on responses, even if integrity wasn&apos;t present in the
request. Needed for static imports triggered by JSCore.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestModuleScript): Add integrity to outgoing
requests for top-level modules, if they don&apos;t already have an integrity
attribute.

Canonical link: <a href="https://commits.webkit.org/279096@main">https://commits.webkit.org/279096@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac14677b6ae8b0f78d38e1263020962d1bd09fa4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52353 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31685 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4774 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55627 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3076 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54658 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38222 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2775 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42571 "Found 1 new test failure: webgl/2.0.y/conformance2/textures/webgl_canvas/tex-3d-rgba4-rgba-unsigned_short_4_4_4_4.html (failure)") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1960 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54449 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29392 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45180 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23651 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/26595 "4 new passes 3 flakes") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2455 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1235 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/45702 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48485 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2605 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57223 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/51861 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27479 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/2642 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49963 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28712 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45300 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49209 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11462 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/29624 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64169 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28457 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12149 "Found 7 new JSC stress test failures: stress/proxy-set-prototype-of.js.bytecode-cache, stress/proxy-set-prototype-of.js.mini-mode, stress/sampling-profiler-display-name.js.dfg-eager-no-cjit-validate, stress/spread-non-array.js.no-llint, wasm.yaml/wasm/references/is_null.js.wasm-eager, wasm.yaml/wasm/stress/immutable-globals.js.wasm-eager, wasm.yaml/wasm/stress/wasm-to-wasm.js.wasm-eager (failure)") | 
<!--EWS-Status-Bubble-End-->

yoavweiss commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -225,10 +224,40 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+        for (auto& [key, value] : *integrityMap) {
+            URL integrityURL = resolve(key, baseURL);
+            if (UNLIKELY(integrityURL.isNull())) {
+                // XXX - here we want to print out an console error without returning - "Integrity URL is not a valid URL starting with '/', './' or '../'"
```
I'd love thoughts/pointers on the best way to do that

marcoscaceres replied:
If you can access the document, maybe:

```c++
document->addConsoleMessage(MessageSource::Security, MessageLevel::Warning, "Integrity URL is not a valid URL starting with '/', './' or '../'"_s);
```

yoavweiss replied:
I don't think we can do that from inside JSCore. But I think I now have a better understanding of how to send those strings to bindings

yoavweiss commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/nonimport-integrity-expected.txt:
```diff
@@ -0,0 +1,16 @@
+FAIL Modulepreload was not loaded as its integrity check was not ignored assert_unreached: Should have rejected: undefined Reached unreachable code
+PASS Modulepreload was loaded as its correct integrity attribute was not ignored
+PASS Modulepreload was loaded as its empty integrity attribute was not ignored
+FAIL Modulepreload was not loaded as its bad integrity attribute was not ignored assert_unreached: Should have rejected: undefined Reached unreachable code
```
This fails because module preload integrity is [not supported](https://bugs.webkit.org/show_bug.cgi?id=238206).

yoavweiss commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https-expected.txt:
```diff
@@ -29,6 +29,10 @@ PASS Script load (url:https://localhost:9443/service-workers/service-worker/reso
 PASS Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test28)
 PASS Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test29)
+PASS Module Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test_module)
+FAIL Module Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test_modulepreload) assert_equals: integrity of Module Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test_modulepreload) must be sha384-foobar. expected "sha384-foobar" but got ""
```
This fails because of modulepreload integrity is not supported in this PR.

yoavweiss commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https-expected.txt:
```diff
@@ -29,6 +29,10 @@ PASS Script load (url:https://localhost:9443/service-workers/service-worker/reso
 PASS Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test29)
+PASS Module Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test_module)
+FAIL Module Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test_modulepreload) assert_equals: integrity of Module Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test_modulepreload) must be sha384-foobar. expected "sha384-foobar" but got ""
+FAIL Module Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test_moduleimport) assert_equals: integrity of Module Script load (url:https://localhost:9443/service-workers/service-worker/resources/sample?test_moduleimport) must be sha384-foobar. expected "sha384-foobar" but got ""
```
This fails because of the way I implemented static import integrity (only on the response side)

yoavweiss commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https.html:
```diff
@@ -90,7 +90,67 @@
       destination: 'script',
       message: `Script load (url:${actual_url})`
     };
-  frame.contentWindow.load_script_with_integrity(actual_url, integrity);
```
This is a manual import, as importing the entire service-workers directory introduced too much unrelated noise

yoavweiss commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -495,14 +496,16 @@ void ScriptModuleLoader::notifyFinished(ModuleScriptLoader& moduleScriptLoader,
-                    rejectWithFetchError(*m_context, WTFMove(promise), ExceptionCode::TypeError, "Cannot load script due to integrity mismatch"_s);
-                    return;
-                }
+        String integrity = downcast<Document>(*m_context).globalObject()->importMap().getIntegrity(sourceURL);
```
This enforces integrity checks on the response even in cases where the integrity value is not propagated through the request (static imports).

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity-valid.html:
```diff
@@ -0,0 +1,37 @@
+  assert_equals(log.length, 1);
+  assert_equals(log[0], "log:NoReferencingScriptValidCheck");
+}, "Script was loaded as its valid integrity check passed");
+</script>
```
Why not just (I haven't tested this... just for simplicity): 

```suggestion
</head>
<body>
<img>
<script>
promise_test(async () => {
  const img = document.querySelector("img");
  await new Promise((resolve) => {
    img.onload = resolve;
    img.src = "/images/green.png?2";
  });
  await import('./resources/log.js?pipe=sub&name=NoReferencingScriptValidCheck');
  assert_equals(log.length, 1);
  assert_equals(log[0], "log:NoReferencingScriptValidCheck");
}, "Script was loaded as its valid integrity check passed");
</script>
```

yoavweiss replied:
The test is testing that the import statement as part of an HTML-based onload event is working as expected (and has the right integrity metadata). So I don't think we want to move the import outside of the HTML based onload handler.

marcoscaceres replied:
ah! I see now. ok, makes sense... Maybe cleaner as?:

```JS
promise_test(async () => {
  const promise = new Promise((resolve, reject) => {
    document.getElementById('test-img').onload = () => {
      import('./resources/log.js?pipe=sub&name=NoReferencingScriptValidCheck').then(resolve).catch(reject);
    };
  });

  await promise;
  assert_equals(log.length, 1);
  assert_equals(log[0], "log:NoReferencingScriptValidCheck");
}, "Script was loaded as its valid integrity check passed");
```

yoavweiss replied:
This is testing the part of [HostLoadImportedModule](https://html.spec.whatwg.org/multipage/webappapis.html#hostloadimportedmodule) where step 6's condition is false and `referencingScript` remains null. For that (IIUC), the onload event must be defined as an HTML attribute, outside of any script tag.

yoavweiss replied:
I'll add a comment to that effect, as it's not necessarily clear

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity.html:
```diff
@@ -0,0 +1,34 @@
+}
+</script>
+<script>
+let promiseResolve;
```
As above... gutting the promise seems unnecessary?

yoavweiss replied:
Same. This test structure is there to guarantee that the import happens as part of an HTML-based event

marcoscaceres replied:
As above... just for improved legibility with the same functionality?

yoavweiss replied:
See above. Added a comment

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/nonimport-integrity.html:
```diff
@@ -0,0 +1,161 @@
+  script.src = "./resources/log.js?pipe=sub&name=ModuleNoIntegrity";
+  const promise = new Promise((resolve, reject) => {
+    script.onload = resolve;
+    script.onerror = () => { reject(Error()); };
```
Maybe add a custom error message here... might make it easier to debug if it fails.

marcoscaceres replied:
ah, no you are forcing rejection... Why not get it to resolve with the error instead... plus you really want the error event, right?

yoavweiss replied:
Adopted suggestion as a helper function (to avoid repeating the same code 3 times..) Thanks!

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/nonimport-integrity.html:
```diff
@@ -0,0 +1,161 @@
+    script.onerror = () => { reject(Error()); };
+  });
+  document.head.appendChild(script);
+  await promise_rejects_js(t, Error, promise);
```
Maybe consider doing it this way? 

```suggestion
  const promise = new Promise((resolve, reject) => {
    script.onload = () = >  {
       reject(new Error("It shouldn't have loaded")) 
     };
    script.onerror = resolve;
  });
  document.head.appendChild(script);
  const event = await promise;
  assert_equals(event.type, "error");
```

yoavweiss replied:
Done!

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/nonimport-integrity.html:
```diff
@@ -0,0 +1,161 @@
+    link.onerror = () => { reject(Error()); };
+  });
+  document.head.appendChild(link);
+  await promise_rejects_js(t, Error, promise);
```
Maybe same as above... resolve with the error event?

yoavweiss replied:
Done!

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/nonimport-integrity.html:
```diff
@@ -0,0 +1,161 @@
+  script.src = "./resources/log.js?pipe=sub&name=ModuleBadIntegrityAttribute";
+  const promise = new Promise((resolve, reject) => {
+    script.onload = resolve;
+    script.onerror = () => { reject(Error()); };
```
maybe same here...

yoavweiss replied:
Done!

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/static-integrity.html:
```diff
@@ -0,0 +1,68 @@
+import './resources/log.js?pipe=sub&name=A';
+</script>
+<script type="module">
+test(t => {
```
nit: whitespace

marcoscaceres replied:
hhmmm.. if these run serially and sync, would it make more sense to write each as:

```HTML
<script>
log = [];
test(t => {...});
</script>
<script>
log = [];
test(t => {...});
</script>
```

marcoscaceres replied:
oh, no, never mind. I see now.

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https.html:
```diff
@@ -90,7 +90,67 @@
       message: `Script load (url:${actual_url})`
     };
-  frame.contentWindow.load_script_with_integrity(actual_url, integrity);
+  frame.contentWindow.load_script_with_integrity(actual_url, integrity, /*type=*/null);
```
Passing null is a bit confusing here, because null has no meaning in the context of load_script_with_integrity(), which just does a falsy check. 

```suggestion
  frame.contentWindow.load_script_with_integrity(actual_url, integrity);
```

yoavweiss replied:
Done

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/fetch-request-resources-iframe.https.html:
```diff
@@ -50,13 +60,42 @@
 }
 
-function load_script_with_integrity(url, integrity) {
+function load_script_with_integrity(url, integrity, type) {
```
if both script and type can be optional, maybe consider using a dictionary here:
 
```
function  load_script_with_integrity(url, { integrity, type } = { }) {
```

And then having these passed as objects. It will be more extensible also if we need to set more attributes in the future.

yoavweiss replied:
Done!

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/dynamic-integrity.html:
```diff
@@ -0,0 +1,87 @@
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script>
+let log;
```
is there anything preventing the `let log` being defined on line 32? Then we can get rid of this script tag.

yoavweiss replied:
Done

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/dynamic-integrity.html:
```diff
@@ -0,0 +1,87 @@
+  promise_test(async t => {
+    log = [];
+    const promise = import(url);
+    await promise_rejects_js(t, TypeError, promise);
```
Maybe just inline: 
```suggestion
    await promise_rejects_js(t, TypeError, import(url));
```

yoavweiss replied:
Done

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/dynamic-integrity.html:
```diff
@@ -0,0 +1,87 @@
+
+test_loaded("./resources/log.js?pipe=sub&name=Bare",["log:Bare"],
+  'script was loaded, as its integrity check was ignored, as it was defined' +
+  ' using a URL that looks like a bare specifier');
```
Nit - can we have this instead? The line breaks/whitepace could end up being a hassle to manually maintain. 

```suggestion
test_not_loaded(
  "./resources/log.js?pipe=sub&name=ResolvesToBadHash",
  "script was not loaded, as its resolved URL failed its integrity check"
);

test_loaded(
  "./resources/log.js?pipe=sub&name=ResolvesToNoHash",
  ["log:NoHash"],
  "script was loaded, as its resolved URL had no integrity check, despite its specifier having one"
);

test_loaded(
  "./resources/log.js?pipe=sub&name=GoodHash",
  ["log:GoodHash"],
  "script was loaded, as its integrity check passed"
);

test_not_loaded(
  "./resources/log.js?pipe=sub&name=BadHashWithNoImport",
  "Script with no import definition was not loaded, as it failed its integrity check"
);

test_not_loaded(
  "bare",
  "Bare specifier script was not loaded, as it failed its integrity check"
);

test_loaded(
  "bare2",
  ["log:F"],
  "Bare specifier used for integrity loaded, as its definition should have used the URL"
);

test_loaded(
  "./resources/log.js?pipe=sub&name=InvalidExtra",
  ["log:InvalidExtra"],
  "script was loaded, as its integrity check passed, despite having an extra invalid hash"
);

test_loaded(
  "./resources/log.js?pipe=sub&name=Suffix",
  ["log:Suffix"],
  "script was loaded, as its integrity check passed, despite having an invalid suffix"
);

test_loaded(
  "./resources/log.js?pipe=sub&name=Multiple",
  ["log:Multiple"],
  "script was loaded, as its integrity check passed given multiple hashes. This also makes sure that the larger hash is picked"
);

test_loaded(
  "./resources/log.js?pipe=sub&name=Bare",
  ["log:Bare"],
  "script was loaded, as its integrity check was ignored, as it was defined using a URL that looks like a bare specifier"
);
```

yoavweiss replied:
Done

marcoscaceres commented on LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity-valid.html:
```diff
@@ -0,0 +1,42 @@
+}
+</script>
+<script>
+let promiseResolve;
```
I'm still not loving this approach... putting the handler declaratively in the HTML makes it quite difficult to debug if things go wrong. Plus the inverted promise seems to add unnecessary complexity. 

Could the following work instead? it seems to achieve the same thing:

If we make the image just `<img>` and then:

```JS
promise_test(async () => {
  const img = document.querySelector("img");
  const promise = new Promise((resolve, reject) => {
    img.onload = () => {
      import('./resources/log.js?pipe=sub&name=NoReferencingScriptValidCheck').then(resolve).catch(reject);
    };
    img.src = "/images/green.png?2";
  });

  await promise;
  assert_equals(log.length, 1);
  assert_equals(log[0], "log:NoReferencingScriptValidCheck");
}, "Script was loaded as its valid integrity check passed");
```

yoavweiss replied:
Having the handler in HTML is the point of the test. See the comment below

marcoscaceres replied:
Even so, shouldn't it then include both the inline HTML handler and the IDL event handler attribute test?

yoavweiss replied:
Sure, I can add such tests to dynamic-integrity (since an IDL event handler test would have a referencing script).

rniwa commented on Source/JavaScriptCore/runtime/ImportMap.h:
```diff
@@ -58,13 +59,17 @@ class ImportMap final : public RefCounted<ImportMap> {
     bool isAcquiringImportMaps() const { return m_isAcquiringImportMaps; }
     void setAcquiringImportMaps() { m_isAcquiringImportMaps = false; }
 
+    JS_EXPORT_PRIVATE String getIntegrity(URL) const;
```
"get" prefix should only be used for a function which takes an out argument.
Since this function returns the result, it shouldn't have "get" prefix. How about `integrityForURL`?
Also, it seems like we should use const URL& instead of passing URL by the value?

yoavweiss replied:
renamed and converted to const URL&

rniwa commented on Source/JavaScriptCore/runtime/ImportMap.h:
```diff
@@ -43,6 +43,7 @@ class ImportMap final : public RefCounted<ImportMap> {
         SpecifierMap m_map;
     };
     using Scopes = Vector<ScopeEntry>;
+    using Integrity = HashMap<URL, String>;
```
I think it's confusing for this hash map to be called `Integrity`. Maybe `IntegrityMap`?

yoavweiss replied:
Renamed!

rniwa commented on Source/JavaScriptCore/runtime/ImportMap.cpp:
```diff
@@ -225,10 +224,47 @@ Expected<void, String> ImportMap::registerImportMap(const SourceCode& sourceCode
+            if (UNLIKELY(integrityURL.isNull())) {
+                errorMessage.append("Integrity URL "_s);
+                errorMessage.append(key);
+                errorMessage.append(" is not a valid URL starting with '/', './' or '../'\n"_s);
```
Isn't this error message a bit misleading given `resolve` also seems to accept a full URL?

yoavweiss replied:
Added a mention of absolute URLs, thanks! :)

rniwa commented on Source/WebCore/dom/ScriptElement.cpp:
```diff
@@ -346,7 +346,10 @@ bool ScriptElement::requestModuleScript(const TextPosition& scriptStartPosition)
-        Ref script = LoadableModuleScript::create(nonce, element->attributeWithoutSynchronization(HTMLNames::integrityAttr), referrerPolicy(), fetchPriorityHint(), crossOriginMode,
+        AtomString integrity = element->attributeWithoutSynchronization(HTMLNames::integrityAttr);
+        if (integrity.isNull())
+            integrity = AtomString(document->globalObject()->importMap().getIntegrity(moduleScriptRootURL));
```
Use { ~ } instead of ( ~ )?

yoavweiss replied:
Done!

yoavweiss reviewed (commented):
Thanks for reviewing!

marcoscaceres reviewed (approved):
Non-reviewer r+ on the tests, with that additional one I suggested.

yoavweiss committed:
Add support for importmap integrity
https://bugs.webkit.org/show_bug.cgi?id=272884

Reviewed by Ryosuke Niwa.

Imported ES modules can't currently have integrity checks, which means
they can't be used in sites where integrity checks are a necessity, for
security and privacy reasons.
This implements such support, by adding an "integrity" section to import
maps.

See https://github.com/whatwg/html/pull/10269

* LayoutTests/TestExpectations: Ignored console logs to avoid flakiness
* LayoutTests/imported/w3c/web-platform-tests/import-maps/WEB_FEATURES.yml: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/data-driven/resources/test-helper.js:
(createTestIframe): Updated through import.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/dynamic-integrity-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/dynamic-integrity.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity-valid-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity-valid.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/no-referencing-script-integrity.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/nonimport-integrity-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/nonimport-integrity.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/static-integrity-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/static-integrity.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/import-maps/w3c-import.log: Imports.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https-expected.txt: Updated.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https.html: Updated to cover Request.integrity.
* LayoutTests/imported/w3c/web-platform-tests/service-workers/service-worker/resources/fetch-request-resources-iframe.https.html: Updated to cover Request.integrity.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/service-workers/service-worker/fetch-request-resources.https-expected.txt: Updated.
* Source/JavaScriptCore/runtime/ImportMap.cpp:
(JSC::ImportMap::resolveImportMatch): Typos and spec link.
(JSC::parseURLLikeModuleSpecifier): Typos and spec link.
(JSC::ImportMap::resolve const): Typos and spec link.
(JSC::normalizeSpecifierKey): Typos and spec link.
(JSC::sortAndNormalizeSpecifierMap): Typos and spec link.
(JSC::ImportMap::registerImportMap): Add parsing for the integrity
section.
(JSC::ImportMap::getIntegrity const): Getter for integrity based on URL.
* Source/JavaScriptCore/runtime/ImportMap.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::importModule): Add integrity to outgoing
requests.
(WebCore::ScriptModuleLoader::notifyFinished): Enforce integrity from
the importmap on responses, even if integrity wasn't present in the
request. Needed for static imports triggered by JSCore.
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestModuleScript): Add integrity to outgoing
requests for top-level modules, if they don't already have an integrity
attribute.

Canonical link: https://commits.webkit.org/279096@main
